// neon-init/dist/index.js
import { detectAvailableEditors } from "./lib/editors.js";
import { usesExtension } from "./lib/extension.js";
import { installNeon } from "./lib/install.js";
import { installAgentSkills } from "./lib/skills.js";
import { confirm, intro, isCancel, log, multiselect, note, outro } from "@clack/prompts";
import { bold, cyan } from "yoctocolors";

//#region src/index.ts
/**
* Initialize Neon projects with MCP Server
*/
async function init() {
	intro("Adding Neon to your project");
	const homeDir = process.env.HOME || process.env.USERPROFILE;
	if (!homeDir) {
		log.error("Could not determine home directory");
		outro("📣 Is this unexpected? Email us at feedback@neon.tech");
		process.exit(1);
	}
	const workspaceDir = process.cwd();
	const availableEditors = await detectAvailableEditors(homeDir);
	if (availableEditors.length === 0) {
		log.warn("No supported editors detected on your system.");
		log.info("Supported editors:");
		log.info("  • VS Code (with Neon Local Connect extension)");
		log.info("  • Cursor (with Neon Local Connect extension)");
		log.info("  • Claude CLI (with MCP Server)");
		const continueAnyway = await confirm({
			message: "Would you like to configure Neon anyway? (You can manually select your editor)",
			initialValue: true
		});
		if (isCancel(continueAnyway) || !continueAnyway) {
			outro("Installation cancelled");
			process.exit(0);
		}
	}
	const response = await multiselect({
		message: "Which editor(s) would you like to configure? (Space to toggle each option, Enter to confirm your selection)",
		options: [
			"Cursor",
			"VS Code",
			"Claude CLI"
		].map((editor) => ({
			value: editor,
			label: editor,
			hint: editor === "Claude CLI" ? "MCP Server" : "Neon Local Connect extension"
		})),
		initialValues: availableEditors,
		required: true
	});
	if (isCancel(response)) {
		outro("Installation cancelled");
		process.exit(0);
	}
	const selectedEditors = response;
	if (selectedEditors.length === 0) {
		log.warn("No editors selected.");
		outro("Installation cancelled");
		process.exit(0);
	}
	const results = await installNeon(homeDir, workspaceDir, selectedEditors);
	const successful = [];
	const failed = [];
	for (const [editor, status] of results.entries()) if (status === "success") successful.push(editor);
	else failed.push(editor);
	if (successful.length > 0) await installAgentSkills(successful);
	const extensionEditors = successful.filter(usesExtension);
	const mcpEditors = successful.filter((e) => !usesExtension(e));
	const failedExtensionEditors = failed.filter(usesExtension);
	const failedMcpEditors = failed.filter((e) => !usesExtension(e));
	if (extensionEditors.length > 0) {
		const extSuccessList = extensionEditors.join(" / ");
		log.step(`Neon Local Connect extension installed for ${extSuccessList}.\n`);
	}
	if (mcpEditors.length > 0) {
		const mcpSuccessList = mcpEditors.join(" / ");
		log.step(`Neon MCP Server is now ready to use with ${mcpSuccessList}.\n`);
	}
	if (failedExtensionEditors.length > 0) {
		log.info("Failed to install extension. For the best local development experience, install Neon Local Connect manually:");
		for (const editor of failedExtensionEditors) if (editor === "VS Code") log.info("  • VS Code: https://marketplace.visualstudio.com/items?itemName=databricks.neon-local-connect");
		else if (editor === "Cursor") log.info("  • Cursor: https://open-vsx.org/extension/databricks/neon-local-connect");
	}
	if (failedMcpEditors.length > 0) log.error(`Failed to configure MCP Server for ${failedMcpEditors.join(" / ")}`);
	if (successful.length === 0) {
		outro("Installation cancelled or failed. Please check the output above and try again.");
		process.exit(1);
	}
	if (extensionEditors.length > 0 && mcpEditors.length === 0) note(`\x1b[0mRestart ${extensionEditors.join(" / ")}, open the Neon extension and type in "${bold(cyan("Get started with Neon"))}\x1b[0m" in your agent chat`, "What's next?");
	else if (mcpEditors.length > 0 && extensionEditors.length === 0) note(`\x1b[0mRestart ${mcpEditors.join(" / ")} and type in "${bold(cyan("Get started with Neon"))}\x1b[0m" in the chat`, "What's next?");
	else note(`\x1b[0mFor ${extensionEditors.join(" / ")}: Restart, open the Neon extension and type in "${bold(cyan("Get started with Neon"))}\x1b[0m" in your agent chat\n\x1b[0mFor ${mcpEditors.join(" / ")}: Restart and type in "${bold(cyan("Get started with Neon"))}\x1b[0m" in the chat`, "What's next?");
	outro("Have feedback? Email us at feedback@neon.tech");
}

//#endregion
export { init };
