// neon-init@2.20.2/dist/lib/skills.js
import { log, spinner } from "@clack/prompts";
import { execa } from "execa";

//#region src/lib/skills.ts
/**
* Maps editor names to the corresponding agent names used by the skills CLI
*/
function getSkillsAgentName(editor) {
	switch (editor) {
		case "Cursor": return "cursor";
		case "VS Code": return "github-copilot";
		case "Claude CLI": return "claude-code";
		default: return "";
	}
}
/**
* Installs Neon agent skills using Vercel's skills CLI
*/
async function installAgentSkills(selectedEditors) {
	if (selectedEditors.length === 0) return true;
	const skillsSpinner = spinner();
	skillsSpinner.start("Installing agent skills for Neon...");
	let anyFailed = false;
	for (const editor of selectedEditors) {
		const agentName = getSkillsAgentName(editor);
		if (agentName === "") {
			log.error(`Unsupported editor: ${editor}`);
			anyFailed = true;
			continue;
		}
		try {
			await execa("npx", [
				"skills",
				"add",
				"neondatabase/agent-skills",
				"--agent",
				agentName,
				"-y"
			], {
				stdio: "pipe",
				timeout: 1e4,
				env: {
					...process.env,
					DISABLE_TELEMETRY: "1"
				}
			});
		} catch (error) {
			log.error(`Failed to install agent skills for ${editor}: ${error instanceof Error ? error.message : "Unknown error"}`);
			anyFailed = true;
		}
	}
	if (anyFailed) {
		skillsSpinner.stop("Agent skills installation completed with errors");
		log.info("You can manually install skills by running: npx skills add neondatabase/agent-skills");
		return false;
	}
	skillsSpinner.stop("Agent skills installed ✓");
	return true;
}

//#endregion
export { installAgentSkills };
